Single file data loader

sahanediriweera opened the pull request.

sahanediriweera committed:
Update app.py

chxru commented on app.py:
```diff
@@ -2,15 +2,35 @@
 max_iter = 10
 
-X = np.random.rand(N, M)
+def load_data(filename):
+    data = []
+    with open(filename, 'r') as csvfile:
+        reader = csv.reader(csvfile)
+        for row in reader:
+            data.append([float(x) for x in row])
+    return np.array(data)
+
+def count_csv_rows(csv_file):
+    row_count = 0
+    with open(csv_file, 'r') as file:
+        reader = csv.reader(file)
+        for row in reader:
+            row_count += 1
+    return row_count
+
+filename = 'data_1000.csv'
```
You can't run the program if you don't have the CSV file. Gonna be a bad experience in fresh runs. Re write this flow

1. Check if the csv exists
2. If file exists, load it
3. Else, generate the csv, save it in disk and load it

chxru commented on app.py:
```diff
@@ -2,15 +2,35 @@
+            data.append([float(x) for x in row])
+    return np.array(data)
+
+def count_csv_rows(csv_file):
+    row_count = 0
+    with open(csv_file, 'r') as file:
+        reader = csv.reader(file)
+        for row in reader:
+            row_count += 1
+    return row_count
```
Might want to look into this
https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.count.html#pandas-dataframe-count

chxru commented on app.py:
```diff
@@ -20,8 +40,10 @@
+start_row = rank*N_per_process
+end_row = N_per_process*(rank+1)-1
+
+data = np.loadtxt(filename, delimiter=',', skiprows=start_row, max_rows=end_row-start_row+1)
```
Something is smelly here, you've load data in line 31 and loading again here by doing a disk IO.
I'd rather 

1. Get a total row count in CSV. You can get the total count without reading the file as mentioned in here https://stackoverflow.com/a/16108605/6862650
2. Calculate N_per_process, determine from which line to which line each process should read
3. Pass the line range, use pd.read_csv to load csv. You can set starting lines and limit by parameters there

sahanediriweera committed:
made discussed changes

The last discussion in PR#3 was the data was loaded again for sequential because we are supposed to take that measurement for the benchmarking

sahanediriweera commented:
Made necessary changes 

The last discussion in PR#3 was the data was loaded again for sequential because we are supposed to take that measurement for the benchmarking

chxru committed:
Merge branch 'master' into 1st-dataloader

chxru reviewed (approved):
Formatting and further refactoring maybe required but LGTM
